### PR TITLE
mag_control: fuse fake heading during leveling fine alignment

### PR DESCRIPTION
**Pull request by bresch**
I recently centralized the mag reset call in `mag_control.cpp` and disabled the mag fusion during the leveling fine alignment phase (https://github.com/PX4/PX4-Autopilot/commit/a685987082e03cef9263c35c8ec20adc708343fa#diff-4e4b2b3cba938a36a704d2d452e4ca2ae427d40254a8c63a5278b0ded821754cR70). This was in theory the correct way of doing it, except that without any heading fusion while fusing fake position (quasi-stationary alignment), the quaternion variance isn't reducing fast enough.

This PR enables fake "zero innovation" heading fusion during the fine alignment phase to speed up the quaternion convergence.

Also enables fake heading fusion when the mag has been rejected (`mag_fault` flag in case of yaw emergency reset for example) to avoid unconstrained yaw variance growth in case of long static (i.e.: no acceleration) flight, where the heading isn't observable anymore.

Tested on a drone where the fine alignment phase wasn't able to complete anymore (triggered by the `at_rest` flag never being true; due to another unrelated issue).

DeepinScreenshot_select-area_20210722123245

Note that the issue is not showing up on most setups, because when the vehicle is detected "at rest", the fake position fusion has a lower observation noise that allows the quaternion to converge much faster.

### CODE REVIEW

File: src/modules/ekf2/EKF/mag_fusion.cpp [hunk 1]
```diff
@@ -781,6 +781,11 @@ void Ekf::fuseHeading()
 				// Vehicle is at rest so use the last moving prediction as an observation
 				// to prevent the heading from drifting and to enable yaw gyro bias learning
 				// before takeoff.
+
+				if (fabsf(_last_static_yaw) < 1e-6f) {
+					_last_static_yaw = predicted_hdg;
```
Comment (dagar) [inline src/modules/ekf2/EKF/mag_fusion.cpp:786]: Default `_last_static_yaw` to NAN instead?

**Comment by bresch** [comment]
backport to v1.12 (using ECL) https://github.com/PX4/PX4-ECL/pull/1029

File: src/modules/ekf2/EKF/mag_control.cpp [hunk 2]
```diff
@@ -67,8 +50,25 @@ void Ekf::controlMagFusion()
 		_num_bad_flight_yaw_events = 0;
 	}
 
-	if (_control_status.flags.mag_fault || !_control_status.flags.tilt_align) {
+	// When operating without a magnetometer and no other source of yaw aiding is active,
+	// yaw fusion is run selectively to enable yaw gyro bias learning when stationary on
+	// ground and to prevent uncontrolled yaw variance growth
+	// Also fuse zero heading innovation during the leveling fine alignment step to keep the yaw variance low
+	if (_params.mag_fusion_type >= MAG_FUSE_TYPE_NONE
```
Comment (priseborough) [inline src/modules/ekf2/EKF/mag_control.cpp:57]: I'm thinking about what happens when _control_status.flags.ev_yaw or _control_status.flags.gps_yaw. This change fixes the mag use case, but if using EV or GPS yaw don't we end up with the same problem during tilt alignment?

Should we move the fusion of zero innovation yaw measurements into the

 	if (!_control_status.flags.tilt_align) {

conditional branch in Ekf::controlFusionModes()

Comment (bresch) [reply src/modules/ekf2/EKF/mag_control.cpp:57]: @priseborough `gps_yaw` and `ev_yaw` fusions cannot start before tilt alignment complete. This means that the fake fusion will always be executed here. However, I agree that we might want to refactor a bit to make it clearer (we should also refactor `fuseHeading` to extract a `fuseFakeHeading` function). We can do that in a follow-up PR.

Comment (priseborough) [reply src/modules/ekf2/EKF/mag_control.cpp:57]: eg: https://github.com/priseborough/PX4-Autopilot/commit/a14496670aeeb5c25172ec43ad975484e395cddd

Comment (priseborough) [reply src/modules/ekf2/EKF/mag_control.cpp:57]: Yes you're right. However this is a special case that is there to stop covariances blowing out during tilt alignmnet so maybe we should move it out of the mag fusion control path anyway

**Pull request merged**